### PR TITLE
[SimpleFSDP] Add mixed precision training support

### PR DESCRIPTION
**Pull request by ruisizhang123**
This PR adds mixed precision training support for SimpleFSDP, together with this change from PyTorch: https://github.com/pytorch/pytorch/pull/150740. 

**(This is a placeholder for now, and shall be merged only after the PyTorch pr is landed.)** 

#### Convergence on debug model

The loss curves are all benchmarked with `training.seed = 42` on 4 GPUs.

- Fully Shard Mode (['dp_shard'], [4]): FSDP2 and SimpleFSDP's losses are perfectly matched
Screenshot 2025-04-05 at 4 06 04 PM

- Hybrid Shard Mode (['dp_replicate', 'dp_shard'], [2, 2]): FSDP2 and SimpleFSDP's losses are similar

Screenshot 2025-04-07 at 6 15 04 PM

The end-to-end SimpleFSDP mixed precision training integration has been proved to work properly in the PR from this fork: https://github.com/tianyu-l/pytorch_intern24/pull/20.

### CODE REVIEW

**Comment by ruisizhang123** [comment]
> LGTM.
> 
> > Replicate Mode (['dp_replicate'], [4]): FSDP2 and SimpleFSDP's losses are almost the same.
> 
> Under pure replicate, FSDP2 is not used in torchtitan. DDP is used, which doesn't handle mixed precision training right now in torchtitan. So you are comparing mixed precision SimpleFSDP with non-mixed precision DDP.
> 
> You can compare HSDP under FSDP2 and SimpleFSDP. I expect them to yield identical loss.

sg, updated.
